src/BodyComponent.h

```diff
@@ -0,0 +1,212 @@
+		}
+	};
+
+	// Returns (and creates) a type-specific pool.
+	template <typename T>
+	static Pool<T> *GetComponentType()
+	{
+		auto iter = m_componentPools.find(TypeId<T>::Get());
+		if (iter == m_componentPools.end()) {
+			auto *pool = new Pool<T>(m_componentIdx++, TypeId<T>::Get());
+			iter = m_componentPools.emplace(TypeId<T>::Get(), pool).first;
+			m_componentTypes.push_back(pool);
+		}
+
+		return static_cast<Pool<T> *>(iter->second.get());
+	}
+
+	// Returns (if present) the polymorphic interface to component associated with the given index
+	// This differs from the type-ID and is volatile between program restarts
+	static PoolBase *GetComponentType(size_t componentIndex)
+	{
+		assert(componentIndex < m_componentTypes.size());
+		return m_componentTypes[componentIndex];
+	}
+
+	// Returns (if present) the polymorphic interface to the component type associated with the given name
+	// This name can be used for serialization or to interface with Lua
+	static PoolBase *GetComponentType(const std::string &typeName)
+	{
+		auto iter = m_componentNames.find(typeName);
+		if (iter != m_componentNames.end())
+			return iter->second;
+
+		return nullptr;
+	}
+
+	// Register a component type to be queryable from Lua with body:GetComponent()
+	template <typename T>
+	static bool RegisterLuaInterface(std::string typeName)
+	{
+		Pool<T> *pool = GetComponentType<T>();
+		if (pool->luaInterface)
+			return false;
+
+		pool->luaInterface = new LuaInterface<T>(pool);
+		m_componentNames.emplace(typeName, pool);
+		return true;
+	}
+
+	// Register a serializer for the given type.
+	// Multiple serializers can be registered for the given type and used while
+	// loading for backwards compatibility, however only the last-registered
+	// serializer will be used when serializing to JSON
+	template <typename T>
+	static bool RegisterSerializer(std::string typeName)
+	{
+		assert(!m_componentSerializers.count(typeName));
+		Pool<T> *pool = GetComponentType<T>();
+
+		SerializerBase *serial = new Serializer<T>(typeName, pool);
+		pool->serializer = serial;
+
+		m_componentSerializers.emplace(typeName, serial);
+		m_componentNames.emplace(typeName, pool);
+		return true;
+	}
+
+	// Returns a pointer to the registered Serializer instance for a type identified by the given name, or nullptr.
+	// To retrieve the serializer instance for a given type index, use GetComponentType(idx)->serializer
+	// or GetComponentType<T>()->serializer.
+	static SerializerBase *GetSerializer(const std::string &typeName)
+	{
+		auto iter = m_componentSerializers.find(typeName);
+		if (iter != m_componentSerializers.end())
+			return iter->second.get();
+
+		return nullptr;
+	}
+
+private:
+	static std::map<size_t, std::unique_ptr<PoolBase>> m_componentPools;
+	static std::map<std::string, std::unique_ptr<SerializerBase>> m_componentSerializers;
+	static std::map<std::string, PoolBase *> m_componentNames;
+	static std::vector<PoolBase *> m_componentTypes;
```

Comment:
Maybe it's worth explicitly limiting the size of this array to 64 somewhere?

Reply:
Yes - when I move to explicit initialization (before merging this PR) I'll add an assert when trying to register components that ensures we don't go above 64 registered components.

Reply:
Added an assert to prevent more than 64 components from being registered.